src/openrct2-ui/windows/Water.cpp

```diff
@@ -37,182 +37,148 @@ static rct_widget window_water_widgets[] = {
+        gWaterToolLowerCost = MONEY32_UNDEFINED;
     }
-}
 
-static void window_water_mousedown(rct_window* w, rct_widgetindex widgetIndex, rct_widget* widget)
-{
-    switch (widgetIndex)
+    void OnClose() override
     {
-        case WIDX_DECREMENT:
-            // Decrement land tool size
-            gLandToolSize = std::max(MINIMUM_TOOL_SIZE, gLandToolSize - 1);
-
-            // Invalidate the window
-            w->Invalidate();
-            break;
-        case WIDX_INCREMENT:
-            // Increment land tool size
-            gLandToolSize = std::min(MAXIMUM_TOOL_SIZE, gLandToolSize + 1);
-
-            // Invalidate the window
-            w->Invalidate();
-            break;
+        // If the tool wasn't changed, turn tool off
+        if (water_tool_is_active())
+        {
+            tool_cancel();
+        }
     }
-}
-
-static void window_water_textinput(rct_window* w, rct_widgetindex widgetIndex, char* text)
-{
-    int32_t size;
-    char* end;
-
-    if (widgetIndex != WIDX_PREVIEW || text == nullptr)
-        return;
 
-    size = strtol(text, &end, 10);
-    if (*end == '\0')
+    void OnMouseUp(rct_widgetindex widgetIndex) override
     {
-        size = std::max(MINIMUM_TOOL_SIZE, size);
-        size = std::min(MAXIMUM_TOOL_SIZE, size);
-        gLandToolSize = size;
-
-        w->Invalidate();
+        switch (widgetIndex)
+        {
+            case WIDX_CLOSE:
+                Close();
+                break;
+            case WIDX_PREVIEW:
+                InputSize();
+                break;
+        }
     }
-}
 
-static void window_water_inputsize(rct_window* w)
-{
-    TextInputDescriptionArgs[0] = MINIMUM_TOOL_SIZE;
-    TextInputDescriptionArgs[1] = MAXIMUM_TOOL_SIZE;
-    window_text_input_open(w, WIDX_PREVIEW, STR_SELECTION_SIZE, STR_ENTER_SELECTION_SIZE, STR_NONE, STR_NONE, 3);
-}
+    void OnMouseDown(rct_widgetindex widgetIndex) override
+    {
+        switch (widgetIndex)
+        {
+            case WIDX_DECREMENT:
+                // Decrement land tool size
+                gLandToolSize = std::max(MINIMUM_TOOL_SIZE, gLandToolSize - 1);
+
+                // Invalidate the window
+                Invalidate();
+                break;
+            case WIDX_INCREMENT:
+                // Increment land tool size
+                gLandToolSize = std::min(MAXIMUM_TOOL_SIZE, gLandToolSize + 1);
+
+                // Invalidate the window
+                Invalidate();
+                break;
+        }
+    }
 
-/**
- *
- *  rct2: 0x006E6BCE
- */
-static void window_water_update(rct_window* w)
-{
-    // Close window if another tool is open
-    if (!water_tool_is_active())
-        window_close(w);
-}
+    void OnUpdate() override
+    {
+        // Close window if another tool is open
+        if (!water_tool_is_active())
+        {
+            Close();
+        }
+    }
 
-/**
- *
- *  rct2: 0x006E6AB8
- */
-static void window_water_invalidate(rct_window* w)
```

Comment:
Is `window_water_invalidate()` mapped to `OnPrepareDraw()` ? I wasn't sure what to do with this in my scenery window PR.

Reply:
Yes we renamed invalidate to prepare draw to better reflect what it is being used for